rs/recovery/src/steps.rs

```diff
@@ -559,71 +589,98 @@ impl Step for UploadAndRestartStep {
-        replace_state.push_str(&format!(
-            "sudo chmod -R --reference={} {};",
-            ic_state_path, upload_dir
-        ));
-        replace_state.push_str(&format!(
-            "sudo chown -R --reference={} {};",
-            ic_state_path, upload_dir
-        ));
-        replace_state.push_str(&format!("sudo rm -r {};", ic_state_path));
-        replace_state.push_str(&format!("sudo mv {} {};", upload_dir, ic_state_path));
-        replace_state.push_str(&format!(
-            r"sudo find {} -type f -exec chmod a-x {{}} \;;",
-            ic_state_path
-        ));
-        replace_state.push_str(&format!(
-            r"sudo find {} -type f -exec chmod go+r {{}} \;;",
-            ic_state_path
-        ));
-        // Note that on older versions of IC-OS this service does not exist.
-        // So try this operation, but ignore possible failure if service
-        // does not exist on the affected version.
-        replace_state.push_str("(sudo systemctl restart setup-permissions || true);");
-        replace_state.push_str("sudo systemctl start ic-replica;");
-        replace_state.push_str("sudo systemctl status ic-replica;");
-
-        ssh_helper.ssh(replace_state)?;
+            let cmd_set_permissions = Self::cmd_set_permissions(&ic_state_path, &upload_dir);
+            let cmd_replace_state = format!(
+                "sudo rm -r {}; sudo mv {} {};",
+                ic_state_path, upload_dir, ic_state_path
+            );
+
+            info!(self.logger, "Restarting replica...");
+            ssh_helper.ssh(Self::CMD_STOP_REPLICA.to_string())?;
+            ssh_helper.ssh(cmd_set_permissions)?;
+            ssh_helper.ssh(cmd_replace_state)?;
+            ssh_helper.ssh(Self::CMD_RESTART_REPLICA.to_string())?;
+        } else {
+            info!(self.logger, "Stopping replica...");
+            exec_cmd(Command::new("bash").arg("-c").arg(Self::CMD_STOP_REPLICA))?;
+
+            info!(self.logger, "Setting file permissions...");
+            let cmd_set_permissions = Self::cmd_set_permissions(&ic_state_path, &src);
+            exec_cmd(Command::new("bash").arg("-c").arg(cmd_set_permissions))?;
+
+            // For local recoveries we first backup the original state, and
+            // then simply `mv` the new state to the upload directory. No
+            // rsync is needed, and thus no checkpoint copying.
+            let backup_path = format!("{}/{}", self.work_dir.display(), OLD_IC_STATE);
+            info!(self.logger, "Moving original state into {}...", backup_path);
+            let mut cmd_backup_state = Command::new("sudo");
+            cmd_backup_state.arg("mv");
+            cmd_backup_state.arg(&ic_state_path);
+            cmd_backup_state.arg(backup_path);
+            exec_cmd(&mut cmd_backup_state)?;
+
+            info!(self.logger, "Moving state locally...");
+            let mut mv_to_target = Command::new("sudo");
+            mv_to_target.arg("mv");
```

Comment:
Does `mv` actually preserve the permissions here? Or are we simply relying on the `setup-permissions` service to fix the permissions again (which may also be fine)?

Reply:
It doesn't say so in the official docs, but `mv` seems to preserve permissions from what I can gather online. It also does so in my testing (at least if we're staying on the same filesystem). So we should be good.